incremental.go

```diff
@@ -0,0 +1,193 @@
+package feedx
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	"github.com/bsm/bfs"
+)
+
+// IncrmentalProduceFunc returns a ProduceFunc closure around an incremental mod time
+type IncrementalProduceFunc func(time.Time) ProduceFunc
+
+// IncrementalProducer produces a continuous incremental feed.
+type IncrementalProducer struct {
+	*ProducerState
+
+	bucket    bfs.Bucket
+	manifest  *bfs.Object
+	ownBucket bool
+	opt       IncrementalProducerOptions
+
+	ctx  context.Context
+	stop context.CancelFunc
+	ipfn IncrementalProduceFunc
+}
+
+type IncrementalProducerOptions struct {
+	ProducerOptions
+}
+
+func (o *IncrementalProducerOptions) norm(lmfn LastModFunc) {
+	if o.Compression == nil {
+		o.Compression = GZipCompression
+	}
+	if o.Format == nil {
+		o.Format = ProtobufFormat
+	}
+	if o.Interval == 0 {
+		o.Interval = time.Minute
+	}
+	o.LastModCheck = lmfn
+}
+
+// NewIncrementalProducer inits a new incremental feed producer.
+func NewIncrementalProducer(ctx context.Context, bucketURL string, opt *IncrementalProducerOptions, lmfn LastModFunc, ipfn IncrementalProduceFunc) (*IncrementalProducer, error) {
+	bucket, err := bfs.Connect(ctx, bucketURL)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := NewIncrementalProducerForBucket(ctx, bucket, opt, lmfn, ipfn)
+	if err != nil {
+		_ = bucket.Close()
+		return nil, err
+	}
+	p.ownBucket = true
+
+	return p, nil
+}
+
+// NewIncrementalProducerForRemote starts a new incremental feed producer for a bucket.
+func NewIncrementalProducerForBucket(ctx context.Context, bucket bfs.Bucket, opt *IncrementalProducerOptions, lmfn LastModFunc, ipfn IncrementalProduceFunc) (*IncrementalProducer, error) {
+	var o IncrementalProducerOptions
+	if opt != nil {
+		o = IncrementalProducerOptions(*opt)
+	}
+	o.norm(lmfn)
+
+	ctx, stop := context.WithCancel(ctx)
+	p := &IncrementalProducer{
+		bucket:        bucket,
+		manifest:      bfs.NewObjectFromBucket(bucket, "manifest.json"),
+		ctx:           ctx,
+		stop:          stop,
+		opt:           o,
+		ipfn:          ipfn,
+		ProducerState: new(ProducerState),
+	}
+
+	// run initial push
+	if _, err := p.push(); err != nil {
+		_ = p.Close()
+		return nil, err
+	}
+
+	// start continuous loop
+	// use error group
+	go p.loop()
+
+	return p, nil
+}
+
+// Close stops the producer.
+func (p *IncrementalProducer) Close() error {
+	p.stop()
+	if p.ownBucket {
+		return p.bucket.Close()
+	}
+	return p.manifest.Close()
+}
+
+func (p *IncrementalProducer) loop() {
+	ticker := time.NewTicker(p.opt.Interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-ticker.C:
+			state, err := p.push()
+			if p.opt.AfterPush != nil {
+				p.opt.AfterPush(state, err)
+			}
+		}
+	}
+}
+
+func (p *IncrementalProducer) push() (*ProducerPush, error) {
+	start := time.Now()
+	atomic.StoreInt64(&p.lastPush, timestampFromTime(start).Millis())
```

Comment:
this should be a private method on `ProducerState`, e.g. `updateLastPush` or `onPushStart`